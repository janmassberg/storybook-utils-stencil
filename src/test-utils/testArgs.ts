export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: (): void => {
        // Implementation
    },
    onTestEvent: (event: Event): void => {
        // Handle event
        console.log(event.type);
    },
    _slot: "Test children",
};

export const testArgsButton = {
    required: true,
    variant: "primary",
    onClick: (event: Event) => {
        // Handle click event
        console.log(event.target);
    },
    _slot: "Label",
};

export const testArgsComponentGenerator = {
    component: "my-component",
    args: testArgs,
};

export const testCopmonentSlot = `<my-child-component foo-bar="foo">
<my-nested-child-component bar-foo="bar"><span class="test-class">Label</span></my-nested-child-component>
</my-child-component>`;
