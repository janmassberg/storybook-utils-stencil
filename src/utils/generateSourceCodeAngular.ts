import { toKebabCase, toUpperCamelCase, ucFirst } from "./stringUtils";
import { formatJsx } from "./formatJsx";
import { formatHtml } from "./formatHtml";
import { objectToString } from "./objectToString";
import { filterHtmlAttributes, filterEventHandlers, filterJsxProperties, ArgsType } from "./argUtils";

const transformAngularAttributes = (
    attributes: ArgsType
) => {
    return attributes.map(([key, value]) => {
        return `${toKebabCase(key)}="${value}"`;
    });
};

export const generateSourceAngularHtml = (
    component: string,
    args: any
): string => {
    const attribs = filterHtmlAttributes(args);
    const props = filterJsxProperties(args);
    const events = filterEventHandlers(args);
    const slot = args._children || "";

    const angularAttributes = transformAngularAttributes(attribs);
    const angularProps = props.map(([key]) => `[${key}]="${key}"\n`);
    const angularEvents = events.map(
        ([key]) => `(${key})="on${ucFirst(key)}($event)"\n`
    );

    const angularComponentSource = `
        <${component} ${angularAttributes.join(" ")} ${angularProps.join(
        " "
    )} ${angularEvents.join(" ")}>${slot}</${component}>`;
    return formatHtml(angularComponentSource);
};

export const generateSourceAngularComponent = (
    component: string,
    args: any
) => {
    const props = filterJsxProperties(args);
    const events = filterEventHandlers(args);
    const angularProps = props.map(([key, value]) => {
        return `public ${key} = ${objectToString(value)};\n`;
    });
    const angularEventHandlers = events.map(
        ([key, value]) => `on${ucFirst(key)} (event: Event) {
            ${value
                .toString()
                .replace(/^[^{]+{/, "")
                .replace(/}$/, "")}
        }`
    );

    const componentName = `${toUpperCamelCase(component)}Example`;
    const componentSelector = `app-${toKebabCase(componentName)}`;
    const templateUrl = `./${toKebabCase(componentName)}.component.html`;
    const angularComponentSource = `import { Component } from "@angular/core";
${" "}
@Component({
    selector: '${componentSelector}',
    templateUrl: '${templateUrl}',
})
export class ${componentName}Component {
    ${angularProps.join("\n")}
    ${angularEventHandlers.join("\n")}
}
    `;
    return formatJsx(angularComponentSource);
};
