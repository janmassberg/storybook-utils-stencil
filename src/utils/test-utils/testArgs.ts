export const testArgs = {
    testNull: null,
    testUndefined: undefined,
    testFalse: false,
    testTrue: true,
    testString: "test-string",
    testNumber: 12345,
    testArray: [1, 2, 3],
    testObject: {
        foo: "bar",
    },
    testFunction: () => {},
    onTestEvent: () => {},
    _children: "Test children",
};
