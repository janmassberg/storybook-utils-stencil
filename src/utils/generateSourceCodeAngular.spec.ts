import { generateSourceAngularComponent, generateSourceAngularHtml } from "./generateSourceCodeAngular";
import { testArgs } from "../test-utils/testArgs";

describe("generateSourceCodeAngular", () => {
    it("should generate valid formatted angular component source code", () => {
        expect(
            generateSourceAngularComponent("my-component", testArgs)
        ).toMatchSnapshot();
    });
    it("should generate valid formatted angular html template source code", () => {
        expect(
            generateSourceAngularHtml("my-component", testArgs)
        ).toMatchSnapshot();
    });
    it("should generate valid formatted angular component source code with children", () => {
        expect(
            generateSourceAngularComponent("my-component", {
                _children: `<my-child-component foo-bar="foo">
    <my-nested-child-component bar-foo="bar">
    <span class="test-class">Test children</span></my-nested-child-component>
</my-child-component>`,
            })
        ).toMatchSnapshot();
    });
    it("should generate valid formatted angular html template source code with children", () => {
        expect(
            generateSourceAngularHtml("my-component", {
                _children: `<my-child-component foo-bar="foo">
    <my-nested-child-component bar-foo="bar">
    <span class="test-class">Test children</span></my-nested-child-component>
</my-child-component>`,
            })
        ).toMatchSnapshot();
    });
});
