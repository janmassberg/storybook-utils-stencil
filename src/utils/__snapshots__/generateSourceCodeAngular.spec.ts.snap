// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSourceCodeAngular should generate valid formatted angular component source code 1`] = `
"import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"app-my-component-example\\",
  templateUrl: \\"./my-component-example.component.html\\",
})
export class MyComponentExampleComponent {
  public testArray = [1, 2, 3];

  public testObject = { foo: \\"bar\\" };

  public testFunction = function () {
    // do nothing
  };

  onOnTestEvent(event: Event) {
    console.log(event.type);
  }
}
"
`;

exports[`generateSourceCodeAngular should generate valid formatted angular component source code with children 1`] = `
"import { Component } from \\"@angular/core\\";

@Component({
  selector: \\"app-my-component-example\\",
  templateUrl: \\"./my-component-example.component.html\\",
})
export class MyComponentExampleComponent {}
"
`;

exports[`generateSourceCodeAngular should generate valid formatted angular html template source code 1`] = `
"<my-component
  test-false=\\"false\\"
  test-true=\\"true\\"
  test-string=\\"test-string\\"
  test-number=\\"12345\\"
  [testArray]=\\"testArray\\"
  [testObject]=\\"testObject\\"
  [testFunction]=\\"testFunction\\"
  (onTestEvent)=\\"onOnTestEvent($event)\\"
>
  Test children
</my-component>
"
`;

exports[`generateSourceCodeAngular should generate valid formatted angular html template source code with children 1`] = `
"<my-component>
  <my-child-component foo-bar=\\"foo\\">
    <my-nested-child-component bar-foo=\\"bar\\">
      <span class=\\"test-class\\">Test children</span>
    </my-nested-child-component>
  </my-child-component>
</my-component>
"
`;
